### PR TITLE
Test `pstr_add_transition_risk()` and `pstr_add_reductions()`

### PR DESCRIPTION
**Pull request by lindadelacombaz**
Dear @maurolepore,

I have created tests for pstr_add_transition_risk() and pstr_add_reductions(). I have also created the helpers file called R/pstr_utils.R which contains the toy data set.

I noticed a few errors thanks to the tests that need to be fixed / were fixed : 
- I wrote that after using the pstr_add_reductions() function, the weo_product_mapper and the weo_flow_mapper columns were renamed product and flow, respectively, but that is not the case. I think it would make sense for them to be renamed. I wrote a FIXME addressing that in the R/pstr_add_reductions.R file.
- I fixed an error in pstr_add_reductions() function by testing some edge cases (like 30 and 70) : the inequality was changed to strictly < 70 instead of <=70 (c.f. [Product sector transition risk](https://drive.google.com/drive/folders/1ag4gMgng3zwyOmEGR9j9jKadWaKR9u0k), page 6).

I unfortunately did not had the time to write tests for pstr_aggregate_score(), but I think this is already a start for my draft pull request !

Thanks ! :D

### CODE REVIEW

File: tests/testthat/test-pstr_add_reductions.R [hunk 1]
```diff
@@ -6,3 +6,75 @@ test_that("hasn't changed", {
...
+  expect_false(hasName(pstr_toy_companies(), "weo_product_mapper"))
+  expect_false(hasName(pstr_toy_companies(), "weo_flow_mapper"))
+  expect_false(hasName(pstr_toy_companies(), "publication"))
+  expect_false(hasName(pstr_toy_companies(), "scenario"))
+  expect_false(hasName(pstr_toy_companies(), "region"))
+  expect_false(hasName(pstr_toy_companies(), "category"))
+  expect_false(hasName(pstr_toy_companies(), "unit"))
+  expect_false(hasName(pstr_toy_companies(), "year"))
+  expect_false(hasName(pstr_toy_companies(), "value"))
+  expect_false(hasName(pstr_toy_companies(), "reductions"))
```
Comment (maurolepore) [inline tests/testthat/test-pstr_add_reductions.R:44]: This doesn't seem to belong here but to belong in tests/testthat/test-pstr_utils.R. You can test the names the toy data must have and then you know it has nothing else than that.

File: tests/testthat/test-pstr_add_reductions.R [hunk 2]
```diff
@@ -6,3 +6,75 @@ test_that("hasn't changed", {
...
+
+  out <- pstr_add_reductions(pstr_toy_companies(), pstr_toy_ep_weo(), pstr_toy_weo_2022())
+  new_columns <- setdiff(names(out), names(pstr_toy_companies()))
+  expect_equal(
+    new_columns, c(
+      "EP_categories_id", "EP_group", "weo_product_mapper",
+      "weo_flow_mapper", "publication", "scenario", "region",
+      "category", "unit", "year", "value", "reductions"
+    )
```
Comment (maurolepore) [inline tests/testthat/test-pstr_add_reductions.R:53]: Nit: This might be better alligned vertically. The UPPERCASE names will almost certainly change. You can make the diff cleaner by putting each column in it's own line.

File: tests/testthat/test-pstr_add_reductions.R [hunk 3]
```diff
@@ -6,3 +6,75 @@ test_that("hasn't changed", {
...
+  expect_true(hasName(out, "y"))
+  expect_true(hasName(out, "z"))
+})
+
+test_that("preserves typeof() input columns", {
+  out <- pstr_add_reductions(pstr_toy_companies(), pstr_toy_ep_weo(), pstr_toy_weo_2022())
+  x <- unname(purrr::map_chr(pstr_toy_companies(), typeof))
+  y <- unname(purrr::map_chr(out[names(pstr_toy_companies())], typeof))
+  expect_equal(x, y)
+})
```
Comment (maurolepore) [inline tests/testthat/test-pstr_add_reductions.R:69]: :1st_place_medal:

File: tests/testthat/test-pstr_add_reductions.R [hunk 4]
```diff
@@ -6,3 +6,75 @@ test_that("hasn't changed", {
...
+
+test_that("outputs 0-rows with an empty data frame as input", {
+  out <- pstr_add_reductions(pstr_toy_companies()[FALSE, ], pstr_toy_ep_weo(), pstr_toy_weo_2022())
```
Comment (maurolepore) [inline tests/testthat/test-pstr_add_reductions.R:72]: Nit: Layout vertically or use the pipe to fit the line in 80 characters.

File: tests/testthat/test-pstr_add_transition_risk.R [hunk 5]
```diff
@@ -8,3 +8,60 @@ test_that("hasn't change", {
...
+    )
+  )
+})
+
+test_that("preserves typeof() input columns", {
+  with_reductions <- pstr_add_reductions(pstr_toy_companies(), pstr_toy_ep_weo(), pstr_toy_weo_2022())
+  out <- pstr_add_transition_risk(with_reductions)
+
+  x <- unname(purrr::map_chr(with_reductions, typeof))
+  y <- unname(purrr::map_chr(out[names(with_reductions)], typeof))
```
Comment (maurolepore) [inline tests/testthat/test-pstr_add_transition_risk.R:52]: Nit: You may import `map_chr()` with `usethis::use_import_from("purrr", "map_chr")`. 
If you use it a few times it's worth importing.

File: tests/testthat/test-pstr_add_transition_risk.R [hunk 6]
```diff
@@ -8,3 +8,60 @@ test_that("hasn't change", {
...
+  out <- pstr_add_transition_risk(with_reductions[FALSE, ])
+  expect_equal(nrow(out), 0L)
+})
+
+test_that("outputs correct values for edge cases", {
+  edge_cases <- pstr_toy_weo_2022(reductions = c(NA, 30, 30.1, 70, 70.1))
+  with_reductions <- pstr_add_reductions(pstr_toy_companies(), pstr_toy_ep_weo(), edge_cases)
+  out <- pstr_add_transition_risk(with_reductions)
+  expect_equal(c("no_sector", "low", "medium", "medium", "high"), out$transition_risk)
+})
```
Comment (maurolepore) [inline tests/testthat/test-pstr_add_transition_risk.R:67]: Awesome! It hurts the brain to think of this cases but it's so important.

**Review by maurolepore — APPROVED** [review]
LGTM

Thanks @lindadelacombaz, your tests cover a great amount of scenarios. You put great effort in thinking how to exercise the code in all possible paths, inlcuding the happy path but also unhappy paths that throw errors, and edge cases. 

Awesome you catch a bug! Thanks for fixing it!

Note I removed your FIXME about renaming columns in `pstr_add_reductions()`. If we need to rename those columns it'll be best to do it elsewhere so that this function stays focused. 

Maybe we need a few `pstr_prepare_*()` (where `*` is the name of a dataset. That function could help us rename columns and other things that ensure the rest of the pipeline works with consistent data. I'll mention that in an issue.

Most of my comments are Nit (things that are nice to have but don't block the PR). You may proceed to resolve each comment (even if choosing to not use my suggestion).

This already leaves the code base in a much better state so it's work merging ASAP. You may continue in a follow up PR.

File: tests/testthat/test-pstr_add_reductions.R [hunk 7]
```diff
@@ -1,8 +1,71 @@
-test_that("hasn't changed", {
-  with_reductions <- pstr_companies |>
-    pstr_add_reductions(pstr_ep_weo, pstr_weo_2022) |>
-    # FIXME: Address warning
-    suppressWarnings()
-  out <- format_robust_snapshot(with_reductions)
-  expect_snapshot(out)
```
Comment (maurolepore) [inline tests/testthat/test-pstr_add_reductions.R:7]: As you wrote more robust and meaningful tests, it's now time to remove the brittle and non-specific snapnshot test. It's then a good idea to run a report of test coverage to confirm all lines are still 100% covered by tests -- which allow us to refactor aggresively knowing that if we break something important then our tests will complain.

image

File: tests/testthat/test-pstr_add_transition_risk.R [hunk 8]
```diff
@@ -1,10 +1,56 @@
-test_that("hasn't change", {
-  with_transition_risk <- pstr_companies |>
-    pstr_add_reductions(pstr_ep_weo, pstr_weo_2022) |>
-    pstr_add_transition_risk() |>
-    # FIXME: Address warning
-    suppressWarnings()
-
-  out <- format_robust_snapshot(with_transition_risk)
-  expect_snapshot(out)
```
Comment (maurolepore) [inline tests/testthat/test-pstr_add_transition_risk.R:9]: Same here. It's best to now remove the snapshot test so the test-coverage report is more useful. If we eventually don't cover a line of code we can focus our attention there.

File: tests/testthat/test-pstr_add_reductions.R [hunk 9]
```diff
@@ -6,3 +6,75 @@ test_that("hasn't changed", {
   out <- format_robust_snapshot(with_reductions)
   expect_snapshot(out)
 })
+
+test_that("throws an error when the user do not give enough parameters", {
+  expect_error(pstr_add_reductions())
+  expect_error(pstr_add_reductions(pstr_toy_companies()))
+  expect_error(pstr_add_reductions(pstr_toy_companies(), pstr_toy_ep_weo()))
+})
```
[the remark this answers is not in the document]
Comment (lindadelacombaz) [reply tests/testthat/test-pstr_add_reductions.R]: Oh right ! I see. Nice.

File: R/pstr_utils.R [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+#' @param EP_sector A string.
+#' @param EP_subsector A string.
+#' @param weo_product_mapper A string.
+#' @param weo_flow_mapper A string.
+#'
+#' @return A toy data set that has the same columns as `pstr_ep_weo`.
+#' @noRd
+pstr_toy_ep_weo <- function(EP_categories_id = 1, EP_group = "a",
+                            EP_sector = "c", EP_subsector = "d",
+                            weo_product_mapper = "e", weo_flow_mapper = "f", ...) {
```
[the remark this answers is not in the document]
Comment (lindadelacombaz) [reply R/pstr_utils.R:38]: Yes, it bugged me too. I'll create an issue :)

File: tests/testthat/test-pstr_add_reductions.R (hunk 7 above)
Comment (lindadelacombaz) [reply tests/testthat/test-pstr_add_reductions.R:7]: :O wow !!
